test/filters/test_grid_filter.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Point CLoud Library (PCL) - www.pointclouds.org
+ * Copyright (c) 2020-, Open Perception
+ *
+ * All rights reserved
+ */
+
+#include <pcl/filters/experimental/grid_filter_base.h>
+#include <pcl/io/pcd_io.h>
+#include <pcl/test/gtest.h>
+#include <pcl/pcl_tests.h>
+#include <pcl/point_types.h>
+
+#include <boost/optional.hpp> // std::optional for C++17
+
+#include <unordered_map>
+
+using namespace pcl;
+using namespace pcl::test;
+
+// Grid structure with necessary declarations for input to GridFilterBase
+
+struct EmptyVoxel {
+  bool voxel_info = false;
+};
+
+// hash map grid
+template <typename PointT>
+struct EmptyMapStruct {
+  using PointCloud = pcl::PointCloud<PointT>;
+  using Grid = std::unordered_map<std::size_t, EmptyVoxel>;
+
+  EmptyMapStruct() : filter_name_("empty_map") {}
+
+protected:
+  // accessing GridFilterBase
+  const experimental::GridFilterBase<EmptyMapStruct>*
+  getDerived()
+  {
+    return static_cast<const experimental::GridFilterBase<EmptyMapStruct>*>(this);
+  }
+
+  bool
+  setUp()
+  {
+    // test passed or failed setUp
+    return getDerived()->getDownsampleAllData();
+  }
+
+  void
+  addPointToGrid(const PointT& pt)
+  {
+    grid_[pt.x].voxel_info = true;
```

Comment:
Should we have some simple hashing function made easily available in PCL?

Reply:
Like [`boost::hash_combine`](https://www.boost.org/doc/libs/1_55_0/doc/html/hash/combine.html) for hashing a point? This has to be carefully implemented due to numerical error.

Do we have any other use case outside grid filters?

Reply:
Will the hash function in cartesian filter be available here? I don't think so, right?

Shall we just make them free functions (with some well documented defaults)?

Reply:
Done